app/Http/Controllers/Webapp/PostController.php

```diff
@@ -0,0 +1,1890 @@
+            if(strlen($uniqueMainKeyword) > 0) {
+                // Insert new main keyword
+                $mainKeyword = new Keywords();
+                $mainKeyword->keyword = $uniqueMainKeyword;
+                $mainKeyword->type = 1;
+                $mainKeyword->save();
+            }
+        }
+
+        $postMetaMkw = PostMeta::getPostMeta($post->id, isset($mainKeyword->id) ? $mainKeyword->id : 0);
+        if(empty($postMetaMkw) && isset($mainKeyword->id)) {
+            // Insert new post-keyword relationship if not exist
+            $postMetaMkw = new PostMeta();
+            $postMetaMkw->post_id = $post->id;
+            $postMetaMkw->keyword_id = $mainKeyword->id;
+            $postMetaMkw->save();
+        }
+
+        // Sub keywords
+        $subs = explode(',', $request->get('subkws'));
+        foreach ($subs as $item) {
+            $trimmedItem = trim($item);
+            if(strlen($trimmedItem) == 0) {
+                continue;
+            }
+
+            $subKeyword = Keywords::getKeyword($trimmedItem, 2);
+
+            if(empty($subKeyword)) {
+                // Insert new keyword
+                $subKeyword = new Keywords();
+                $subKeyword->keyword = $trimmedItem;
+                $subKeyword->type = 2;
+                $subKeyword->save();
+
+                // Save post-keyword relationship
+            }
+
+            $postMetaSkw = PostMeta::getPostMeta($post->id, $subKeyword->id);
+            if(empty($postMetaSkw)) {
+                // Insert new post-keyword relationship if not exist
+                $postMetaSkw = new PostMeta();
+                $postMetaSkw->post_id = $post->id;
+                $postMetaSkw->keyword_id = $subKeyword->id;
+                $postMetaSkw->save();
+            }
+        }
+
+        return [
+            'data' => [
+                'id' => $post->id,
+                'checksum' => $calculatedChecksum
+            ],
+            'next' => action('Webapp\\PostController@getEdit', $post->id)
+        ];
+    }
+
+    public function getPublish(Request $request, $postId = 0) {
+        $this->activityLog->setPostId($postId);
+
+        if(empty($post = Post::find($postId))
+            || $this->currentUser->is(User::CONTRIBUTOR)) {
+            $this->activityLog->setInfo('Not found');
+            return '404 Not Found';
+        }
+
+        $this->activityLog->addMeta('PreviousPublishedStatus', $post->published_status);
+
+        if(!$post->canPublishBy($this->currentUser)) {
+            $this->_flashMessage($request, 'Publish is not permitted', 'danger');
+            $this->activityLog->setInfo('Not permitted');
+            return redirect()->to(action('Webapp\\PostController@getIndex'))->send();
+        }
+
+        if($post->isLock($this->currentSession)) {
+            $this->_flashMessage($request, 'このコンテンツは現在ロックされています。変更できません。', 'danger');
+            $this->activityLog->setInfo('Locked');
+            if(!$request->get('selected_ids'))
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Define a constant instead of duplicating this literal "selected_ids" 12 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://sonar.relipa.blog/coding_rules#rule_key=php%3AS1192)
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Add curly braces around the nested statement(s). [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://sonar.relipa.blog/coding_rules#rule_key=php%3AS121)